graphql_compiler/compiler/emit_sql.py

```diff
@@ -640,6 +653,22 @@ def generate_subquery(self):
         return alias
 
 
+@dataclass
+class ColumnRouter:
+    """Container for columns selected from a variety of selectables.
+
+    ContextFields selecting from locations inside a CTE need to be redirected to get those columns
+    from the corresponding columns exposed by the CTE instead. A ColumnRouter can be used to store
+    these column name mappings.
+
+    The Selectable.c property is the only property of selectables used by ContextFields, so that's
+    the only property this class needs to implement to serve as a Selectable.
+    TODO(bojanserafimov): make an abstract class instead of duck-typing this.
```

Comment:
@obi1kenobi want your opinion on this solution before I go ahead typing.

The alternative is to:
1. Create a `class GeneratedLocation(BaseLocation)`
2. Make a `CreateSubquery` IR block that will be used for CTEs
3. Put all used fields of all locations before the `CreateSubquery` block into the `CreateSubquery` block with some name disambiguation
4. Rewrite all following locations to point to the subquery instead of the locations it's wrapping.

Let me know if you see a cleaner solution.

Reply:
Yeah this seems the best option at the moment. Maybe if/when we re-do the IR we can come up with something better. The alternative approach seems like a heck of a lot of work, only to produce a solution that is still held together by string and duct tape.